### PR TITLE
Zstd compressor dummy

### PR DESCRIPTION
**Pull request by clusty**

### CODE REVIEW

File: src/examples/deepExamples.cpp [hunk 1]
```diff
@@ -147,7 +148,8 @@ writeDeepScanlineFile (
 
     Array2D<half*>& dataA,
 
-    Array2D<unsigned int>& sampleCount)
+    Array2D<unsigned int>& sampleCount,
```
Comment (pleprince) [inline src/examples/deepExamples.cpp:148]: Formatting (using openexr's .clang-format)
```suggestion
    Array2D<unsigned int>& sampleCount,
    Compression            compression = Compression::ZIPS_COMPRESSION)
```

File: src/lib/OpenEXR/CMakeLists.txt [hunk 2]
```diff
@@ -220,3 +222,7 @@ openexr_define_library(OpenEXR
     OpenEXR::IlmThread
     OpenEXR::OpenEXRCore
   )
+
+target_include_directories(OpenEXR PUBLIC "/home/vladal/bin/include/")
```
Comment (pleprince) [inline src/lib/OpenEXR/CMakeLists.txt:226]: How do we fix this ? I will check how deps are generally handled here.

File: src/lib/OpenEXR/ImfCompressionAttribute.cpp [hunk 3]
```diff
@@ -57,7 +57,7 @@ CompressionAttribute::readValueFrom (
         tmp != ZIPS_COMPRESSION && tmp != ZIP_COMPRESSION &&
         tmp != PIZ_COMPRESSION && tmp != PXR24_COMPRESSION &&
         tmp != B44_COMPRESSION && tmp != B44A_COMPRESSION &&
-        tmp != DWAA_COMPRESSION && tmp != DWAB_COMPRESSION)
+        tmp != DWAA_COMPRESSION && tmp != DWAB_COMPRESSION && tmp != ZSTD_COMPRESSION)
```
Comment (pleprince) [inline src/lib/OpenEXR/ImfCompressionAttribute.cpp]: Formatting
```suggestion
        tmp != DWAA_COMPRESSION && tmp != DWAB_COMPRESSION && 
        tmp != ZSTD_COMPRESSION)
```

File: src/lib/OpenEXR/ImfCompressor.cpp [hunk 4]
```diff
@@ -63,6 +64,7 @@ isValidCompression (Compression c)
         case B44_COMPRESSION:
         case B44A_COMPRESSION:
         case DWAA_COMPRESSION:
+        case ZSTD_COMPRESSION:
         case DWAB_COMPRESSION: return true;
```
Comment (pleprince) [inline src/lib/OpenEXR/ImfCompressor.cpp]: Don't separate related schemes,
```suggestion
        case DWAA_COMPRESSION:
        case DWAB_COMPRESSION:
        case ZSTD_COMPRESSION: return true;
```

File: src/lib/OpenEXR/ImfCompressor.cpp [hunk 5]
```diff
@@ -140,7 +143,8 @@ newCompressor (Compression c, size_t maxScanLineSize, const Header& hdr)
                 static_cast<int> (maxScanLineSize),
                 256,
                 DwaCompressor::STATIC_HUFFMAN);
-
+        case ZSTD_COMPRESSION:
+            return new ZstdCompressor (hdr, maxScanLineSize);
```
Comment (pleprince) [inline src/lib/OpenEXR/ImfCompressor.cpp]: Formatting.
```suggestion

        case ZSTD_COMPRESSION:
            return new ZstdCompressor (hdr, maxScanLineSize);

```

File: src/lib/OpenEXR/ImfCompressor.cpp [hunk 6]
```diff
@@ -162,6 +166,7 @@ numLinesInBuffer (Compression comp)
         case B44_COMPRESSION:
         case B44A_COMPRESSION:
         case DWAA_COMPRESSION: return 32;
+        case ZSTD_COMPRESSION: return 1;// FIXME
```
Comment (pleprince) [inline src/lib/OpenEXR/ImfCompressor.cpp]: Why does this needs fixing ?

File: src/lib/OpenEXR/ImfHeader.cpp [hunk 7]
```diff
@@ -70,9 +70,11 @@ struct CompressionRecord
     {
         exr_get_default_zip_compression_level(&zip_level);
         exr_get_default_dwa_compression_quality(&dwa_level);
+        exr_get_default_zstd_compression_level(&zstd_level);
     }
     int   zip_level;
     float dwa_level;
+    int zstd_level;
```
Comment (pleprince) [inline src/lib/OpenEXR/ImfHeader.cpp]: Formatting
```suggestion
    int   zstd_level;
```

File: src/lib/OpenEXR/ImfZstdCompressor.cpp [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+int
+ZSTD_uncompress_impl (const char* inPtr, int inSize, Imf::ZstdCompressor::raw_ptr& outPtr)
+{
+    auto rSize = ZSTD_getFrameContentSize(inPtr, inSize);
+    outPtr = Imf::ZstdCompressor::raw_ptr((char*)malloc (rSize), &free);
+    auto const dSize = ZSTD_decompress(outPtr.get(), rSize, inPtr, inSize);
+    return dSize;
+}
+*/
```
Comment (pleprince) [inline src/lib/OpenEXR/ImfZstdCompressor.cpp]: remove dead code

File: src/lib/OpenEXR/ImfZstdCompressor.cpp [hunk 9]
```diff
@@ -0,0 +1,123 @@
+//
```
Comment (pleprince) [inline src/lib/OpenEXR/ImfZstdCompressor.cpp]: Please run clang-format on this file with `-style=file`.

File: src/lib/OpenEXR/ImfZstdCompressor.h [hunk 10]
```diff
@@ -0,0 +1,29 @@
+#pragma once
```
Comment (pleprince) [inline src/lib/OpenEXR/ImfZstdCompressor.h:1]: Please run clang-format on this file with `-style=file`.

File: src/lib/OpenEXRCore/base.c [hunk 11]
```diff
@@ -208,3 +208,19 @@ exr_get_default_dwa_compression_quality (float* q)
...
 }
+
+static int sDefaultZstdLevel = 5;
+
+void
+exr_set_default_zstd_compression_level (int q)
+{
+    if (q < 0) q = 0;
+    if (q > 9) q = 9;
+    sDefaultDwaLevel = q;
```
Comment (pleprince) [inline src/lib/OpenEXRCore/base.c]: Looks like a bug...
```suggestion
    sDefaultZstdLevel = q;
```

File: src/lib/OpenEXRCore/base.c [hunk 12]
```diff
@@ -208,3 +208,19 @@ exr_get_default_dwa_compression_quality (float* q)
 {
     if (q) *q = sDefaultDwaLevel;
 }
+
+static int sDefaultZstdLevel = 5;
```
Comment (pleprince) [inline src/lib/OpenEXRCore/base.c:213]:
```suggestion
// 9 is 20% more expensive to compress. Decompression time remains constant.
static int sDefaultZstdLevel = 5;
```

**Review by pleprince — CHANGES_REQUESTED** [review]
Mostly formatting issues (always use the project's `.clang-format`).
One bug and an open question about adding BLOSC as a dependency.
If you can do this asap, I will make a second pass. ;-)

File: src/lib/OpenEXR/ImfCompressor.cpp (hunk 6 above)
Comment (clusty) [reply src/lib/OpenEXR/ImfCompressor.cpp]: I am still unclear about single vs multi scanline compression. Will probably assume single for now.

File: src/lib/OpenEXR/ImfZstdCompressor.cpp (hunk 9 above)
Comment (clusty) [reply src/lib/OpenEXR/ImfZstdCompressor.cpp]: done

File: src/lib/OpenEXR/ImfZstdCompressor.h (hunk 10 above)
Comment (clusty) [reply src/lib/OpenEXR/ImfZstdCompressor.h:1]: Done

File: src/lib/OpenEXR/ImfZstdCompressor.cpp (hunk 8 above)
Comment (clusty) [reply src/lib/OpenEXR/ImfZstdCompressor.cpp]: Done

File: src/lib/OpenEXR/ImfZstdCompressor.cpp [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+#include "ImfChannelList.h"
+#include "ImfMisc.h"
+namespace
+{
+class BloscInit
+{
+public:
+    static void Init () { getInstance (); }
+    BloscInit (const BloscInit&) = delete;
+    BloscInit& operator= (const BloscInit&) = delete;
```
Comment (pleprince) [inline src/lib/OpenEXR/ImfZstdCompressor.cpp]: formatting diff on my end:
```suggestion
    BloscInit (const BloscInit&)            = delete;
    BloscInit& operator= (const BloscInit&) = delete;
```

File: src/lib/OpenEXR/ImfZstdCompressor.cpp [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    , _outBuffer (nullptr, &free)
+    , _schunk (nullptr, &blosc2_schunk_free)
+{}
+
+int
+ZstdCompressor::numScanLines () const
+{
+    return _numScanLines; // Needs to be in sync with ImfCompressor::numLinesInBuffer
+}
+int
```
Comment (pleprince) [inline src/lib/OpenEXR/ImfZstdCompressor.cpp]: new line between methods for legibility
```suggestion
}

int
```

File: src/lib/OpenEXR/ImfZstdCompressor.cpp [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+ZstdCompressor::compress (
+    const char* inPtr, int inSize, int minY, const char*& outPtr)
+{
+    auto ret = BLOSC_compress_impl (inPtr, inSize, outPtr);
+    return ret;
+}
+int
```
Comment (pleprince) [inline src/lib/OpenEXR/ImfZstdCompressor.cpp]: same
```suggestion
}

int
```

File: src/lib/OpenEXR/ImfZstdCompressor.cpp [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+ZstdCompressor::BLOSC_compress_impl (
+    const char* inPtr, int inSize, const char*& out)
+{
+    BloscInit::Init ();
+    blosc2_cparams cparams = BLOSC2_CPARAMS_DEFAULTS;
+
+    cparams.typesize = pixelTypeSize (PixelType::FLOAT); // Expect Float values
+    cparams.clevel =
+        header ()
+            .zstdCompressionLevel (); // 9 is about a 20% increase in compression compared to 5. Decompression speed is unchanged.
```
Comment (pleprince) [inline src/lib/OpenEXR/ImfZstdCompressor.cpp]: don't like long comments inlined
```suggestion
    // clevel 9 is about a 20% increase in compression compared to 5.
    // Decompression speed is unchanged.
    cparams.clevel   = header ().zstdCompressionLevel ();
```

**Review by pleprince — APPROVED** [review]
LGTM. Thanks Vlad !

**Review by pleprince — APPROVED** [review]

**Pull request merged**